File: AoC2021/D02/lbm.py
from re import findall
from functools import reduce
from operator import mul

def f1(e1, e2):
    x, y = e1
    d, val = e2
    return                               \
             (x, y - val) if d == 'up'   \
        else (x, y + val) if d == 'down' \
        else (x + val, y)

def f2(e1, e2):
    aim, x, y = e1
    d, val = e2
    return                                    \
             (aim - val, x, y) if d == 'up'   \
        else (aim + val, x, y) if d == 'down' \
        else (aim, x + val, y + aim * val)

inp = [(d, int(val)) for d, val in findall(r'(\w+) (\d+)\n', open('input.txt', 'r').read())]
print('Star 1:', reduce(mul, reduce(f1, inp, (0, 0))))
print('Star 2:', reduce(mul, reduce(f2, inp, (0, 0, 0))[1:]))

File: AoC2021/D02/lbm.hs
type Command = (String, Int)

parse :: String -> [Command]
parse s = map (parse' . words) (lines s)
  where
      parse' [d, val] = (d, read val :: Int)

star1 :: [Command] -> Int
star1 = product . foldr f [0, 0]
  where
      f (d, val) [x, y]
        | d == "up"   = [x, y - val]
        | d == "down" = [x, y + val]
        | otherwise   = [x + val, y]

star2 :: [Command] -> Int
star2 = product . tail . foldl f [0, 0, 0]
  where
      f [aim, x, y] (d, val)
        | d == "up"   = [aim - val, x, y]
        | d == "down" = [aim + val, x, y]
        | otherwise   = [aim, x + val, y + aim * val]

main = do
    inp <- readFile "input.txt"
    let l = parse inp
    putStrLn $ "Star 1: " ++ (show . star1 $ l)
    putStrLn $ "Star 2: " ++ (show . star2 $ l)
